File: lottie/src/main/java/org/thorvg/lottie/LottieAnimationView.java
package org.thorvg.lottie;

import android.content.Context;
import android.content.res.Resources;
import android.content.res.TypedArray;
import android.graphics.Canvas;
import android.graphics.drawable.Drawable;
import android.os.Parcelable;
import android.util.AttributeSet;
import android.view.View;
import androidx.annotation.DrawableRes;
import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import org.thorvg.lottie.LottieDrawable.LottieAnimationListener;

public class LottieAnimationView extends View {

    private LottieDrawable mDrawable;
    private LottieAnimationListener mListener;

    private int mResId = Resources.ID_NULL;

    private boolean mOnAttached = false;

    public LottieAnimationView(Context context) {
        this(context, null);
    }

    public LottieAnimationView(Context context, @Nullable AttributeSet attrs) {
        this(context, attrs, 0);
    }

    public LottieAnimationView(Context context, @Nullable AttributeSet attrs, int defStyleAttr) {
        this(context, attrs, defStyleAttr, 0);
    }

    public LottieAnimationView(Context context, @Nullable AttributeSet attrs, int defStyleAttr,
            int defStyleRes) {
        super(context, attrs, defStyleAttr, defStyleRes);

        final TypedArray a = context.obtainStyledAttributes(
                attrs, R.styleable.LottieAnimationView, defStyleAttr, defStyleRes);
        mResId = a.getResourceId(R.styleable.LottieAnimationView_lottieDrawable, mResId);
        a.recycle();
    }

    public void setLottieDrawableResource(@DrawableRes int resId) {
        if (mResId != resId) {
            mResId = resId;

            if (mDrawable != null) {
                mDrawable.release();
            }

            createLottieDrawable();
        }
    }

    private void createLottieDrawable() {
        if (mOnAttached && mResId != Resources.ID_NULL && mDrawable == null) {
            mDrawable = LottieDrawable.create(getContext().getResources(), mResId);
            if (mDrawable != null) {
                mDrawable.setCallback(this);
                mDrawable.setAnimationListener(mListener);
                if (mDrawable.isAutoPlay()) {
                    startAnimation();
                }
            }
        }
    }

    public void setSize(int width, int height) {
        if (mDrawable != null) {
            mDrawable.setSize(width, height);
        }
    }

    public void startAnimation() {
        if (mDrawable != null) {
            mDrawable.start();
        }
    }

    public void stopAnimation() {
        if (mDrawable != null) {
            mDrawable.stop();
        }
    }

    public void pauseAnimation() {
        if (mDrawable != null) {
            mDrawable.pause();
        }
    }

    public void resumeAnimation() {
        if (mDrawable != null) {
            mDrawable.resume();
        }
    }

    @Override
    protected void onMeasure(int widthMeasureSpec, int heightMeasureSpec) {
        super.onMeasure(widthMeasureSpec, heightMeasureSpec);

        if (mDrawable != null) {
            mDrawable.setSize(getMeasuredWidth(), getMeasuredHeight());
        }
    }

    @Override
    protected void onAttachedToWindow() {
        mOnAttached = true;

        super.onAttachedToWindow();

        createLottieDrawable();
    }

    @Override
    protected void onDetachedFromWindow() {
        mOnAttached = false;

        super.onDetachedFromWindow();

        if (mDrawable != null) {
            mDrawable.release();
            mDrawable = null;
        }
    }

    @Override
    public void draw(Canvas canvas) {
        super.draw(canvas);

        if (mDrawable != null) {
            mDrawable.draw(canvas);
        }
    }

    @Override
    public void invalidateDrawable(@NonNull Drawable drawable) {
        super.invalidateDrawable(drawable);
        invalidate();
    }

    public void setAnimationListener(LottieAnimationListener listener) {
        mListener = listener;
    }

    @Nullable
    @Override
    protected Parcelable onSaveInstanceState() {
        return super.onSaveInstanceState();
    }

    @Override
    protected void onRestoreInstanceState(Parcelable state) {
        super.onRestoreInstanceState(state);
    }
}

File: lottie/src/main/java/org/thorvg/lottie/LottieAnimationView.kt
/*
 * Copyright (c) 2025 - 2026 ThorVG project. All rights reserved.

 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:

 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.

 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */

package org.thorvg.lottie

import android.content.Context
import android.content.res.Resources
import android.graphics.Canvas
import android.graphics.drawable.Drawable
import android.os.Parcelable
import android.util.AttributeSet
import android.view.View
import androidx.annotation.DrawableRes
import org.thorvg.lottie.LottieDrawable.LottieAnimationListener

class LottieAnimationView @JvmOverloads constructor(
    context: Context, attrs: AttributeSet? = null, defStyleAttr: Int = 0,
    defStyleRes: Int = 0
) : View(context, attrs, defStyleAttr, defStyleRes) {
    private var drawable: LottieDrawable? = null
    private var listener: LottieAnimationListener? = null

    private var resId = Resources.ID_NULL

    private var onAttached = false

    init {
        val a = context.obtainStyledAttributes(
            attrs, R.styleable.LottieAnimationView, defStyleAttr, defStyleRes
        )
        resId = a.getResourceId(R.styleable.LottieAnimationView_lottieDrawable, resId)
        a.recycle()
    }

    fun setLottieDrawableResource(@DrawableRes resId: Int) {
        if (this@LottieAnimationView.resId != resId) {
            this@LottieAnimationView.resId = resId

            drawable?.release()

            createLottieDrawable()
        }
    }

    private fun createLottieDrawable() {
        if (onAttached && resId != Resources.ID_NULL && drawable == null) {
            drawable = LottieDrawable.create(context.resources, resId)
            drawable?.callback = this
            drawable?.setAnimationListener(listener)
            if (drawable?.isAutoPlay == true) {
                startAnimation()
            }
        }
    }

    fun setSize(width: Int, height: Int) {
        drawable?.setSize(width, height)
    }

    fun startAnimation() {
        drawable?.start()
    }

    fun stopAnimation() {
        drawable?.stop()
    }

    fun pauseAnimation() {
        drawable?.pause()
    }

    fun resumeAnimation() {
        drawable?.resume()
    }

    override fun onMeasure(widthMeasureSpec: Int, heightMeasureSpec: Int) {
        super.onMeasure(widthMeasureSpec, heightMeasureSpec)

        drawable?.setSize(measuredWidth, measuredHeight)
    }

    override fun onAttachedToWindow() {
        onAttached = true

        super.onAttachedToWindow()

        createLottieDrawable()
    }

    override fun onDetachedFromWindow() {
        onAttached = false

        super.onDetachedFromWindow()

        drawable?.release()
        drawable = null
    }

    override fun draw(canvas: Canvas) {
        super.draw(canvas)

        drawable?.draw(canvas)
    }

    override fun invalidateDrawable(drawable: Drawable) {
        super.invalidateDrawable(drawable)
        invalidate()
    }

    fun setAnimationListener(listener: LottieAnimationListener?) {
        this@LottieAnimationView.listener = listener
    }

    override fun onSaveInstanceState(): Parcelable? {
        return super.onSaveInstanceState()
    }

    override fun onRestoreInstanceState(state: Parcelable?) {
        super.onRestoreInstanceState(state)
    }
}